feat: build wheels using CIBuildWheel

tseaver opened the pull request.

tseaver committed:
ci: try using cibuildwheel

tseaver committed:
ci: consolidate jobs

tseaver committed:
fix: use container-relative install prefix

tseaver committed:
fix: make cibuildwheel run verbosely

tseaver committed:
fix: dead reckoning from install prefix

tseaver committed:
ci: must build library in container

tseaver committed:
ci: use 'wheelhouse' for artifacts

tseaver committed:
ci: test built wheels

tseaver committed:
ci: enable macos wheel builds

tseaver committed:
ci: try building C lib on windows

tseaver committed:
restore linux / macos builds

tseaver committed:
ci: try building from default prefix on Windows

tseaver committed:
ci: try building from default prefix on Windows moar

tseaver committed:
ci: try building from default prefix on Windows harder

tseaver committed:
ci: debug installed prefix on Windows

tseaver committed:
ci: disable pure-Python build under CIBW

tseaver committed:
ci: 'runtime_library_dirs' borks on Windows

tseaver committed:
ci: moar tracer bullitz

tseaver committed:
ci: 'runtime_library_dirs' borks on Windows moar

tseaver committed:
ci: try pre-building extension on Windows

tseaver committed:
ci: debug copying library DLL into project

tseaver committed:
ci: try rationalizing 'copy_dll' dance

tseaver committed:
ci: try static build of C library.

tseaver committed:
ci: go back to explicit install prefix for C library

tseaver committed:
ci: build only AMD64 wheel

We need to work out how the architcture labels play between Github Workflow
and CIBuildWheel, so that we only build Win32 wheels when we've built the
corresponding C library.

tseaver committed:
ci: use platform-specific install prefix for C library

tseaver committed:
ci: tweak setting install prefix

tseaver committed:
ci: tweak archs, install prefix

tseaver committed:
ci: use envvar everywhere

tseaver committed:
ci: use envvar everywhere moar

tseaver committed:
ci: cache C library build

tseaver committed:
ci: don't use platform in lib prefix

tseaver committed:
ci: fix c library cache test

tseaver committed:
ci: tweak CIBW_ARCHS setting

tseaver committed:
ci: restore linux / macos builds

andrewsg commented on .github/workflows/use-cibuildwheel.yml:
```diff
@@ -0,0 +1,167 @@
+      matrix:
+        os:
+        - macos-10.15
+       #- macos-11
```
Due to the changes we made previously to fix the OS X 11 bug, we ended up not needing a separate Big Sur build, correct?

tseaver replied:
I believe so -- even the wheels built on Big Sur were segfaulting on Big Sur with Python 3.8 before that fix.

andrewsg commented on .github/workflows/use-cibuildwheel.yml:
```diff
@@ -0,0 +1,167 @@
+      with:
+        submodules: 'recursive'
+
+    - name: Build C Library
```
Why is this needed as a separate github action step if it's also being built as a cibuildwheel prep step?

tseaver replied:
We only build it as a prep step on Linux, where we need to build it using the same `manylinux` image being used to build the wheels.  Here, splitting it out can help optimize the build (via a cache -- see below where I added the cache for the Windows library build), as well as making the built libraries potentitally available as downloadable artifacts.

andrewsg reviewed (commented):
Shall we change the PR name to feat: so that we get a version bump? I should have added one back in 1.1.3, but better late than never.

tseaver committed:
ci: cache MacOS library build

tseaver commented:
@andrewsg 
> Shall we change the PR name to feat: so that we get a version bump? I should have added one back in 1.1.3, but better late than never.

Makes sense -- the M1 builds are certainly a new feature.

One question before we merge this PR:  the workflow I have now makes more sense for the release-tag build than for a presubmit:  it bulids and tests all the wheels we know how to build.  Perhaps we could trim back the wheels we build and test for the presubmit? (No `Win32`, maybe, and maybe only a couple of `manylinux` versions?

tseaver commented on .github/workflows/use-cibuildwheel.yml:
```diff
@@ -0,0 +1,181 @@
+      uses: pypa/cibuildwheel@v2.1.1
+      env:
+        CIBW_ARCHS_WINDOWS: ${{ steps.platform-arch.outputs.arch }}
+        CIBW_ENVIRONMENT: CRC32C_PURE_PYTHON="0" CRC32C_INSTALL_PREFIX="D:/a/python-crc32c/python-crc32c/usr"
```
I'm not really satisfied that using this hard-wired expansion of the path is best, but the various stabs I made at using `${{ github.workspace }}` to build the path all ended up getting mangled by CIBW passing through backslash-mangled versions.

andrewsg replied:
It's pretty comical that we are still dealing with the same backslash issues after some 40 years of DOS and Windows.

tseaver committed:
ci: test dir-ness after build / cache load

tseaver committed:
ci: try to synthesize forward-slash-only prefix

andrewsg commented:
Normally I'd want to keep the presubmit as close to release as possible. What's the benefit of trimming it back in this case?

tseaver committed:
ci: raw string literal

tseaver committed:
ci: set prefix output var

tseaver committed:
ci: document the workarounds

tseaver commented:
@andrewsg 
> Normally I'd want to keep the presubmit as close to release as possible. What's the benefit of trimming it back in this case?

Turnaround time:  building all the wheels is slow, and seems unlikely to uncover any issues.  Not to mention heating the planet. :)

andrewsg commented:
Okay, if we can cut down on presubmit safely without exposing ourselves to any risk of cutting a release where the build doesn't complete and partial artifacts are pushed out, then I'm fine with that.

tseaver committed:
ci: polish presubmit

tseaver committed:
ci: restore original workflow filename

tseaver committed:
ci: skip PyPy build on presubmit

tseaver committed:
ci: skip universal wheel for presubmit

tseaver commented:
@andrewsg I think we can add handling a [`scheduled` event](https://docs.github.com/en/actions/reference/events-that-trigger-workflows#scheduled-events) to the "build everything" workflow, and have it [skip the "push to PyPI" job](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idif) in that case.  We can then use the artifacts from the nightly build to verify what would be released.

andrewsg commented:
Hmm, that sounds a bit complex... we've got quite a bit going on in this release, do you think we should push that through as well or can we accept the longer presubmits for now?

tseaver committed:
ci: configure GHA job for building all wheels, pushing to PyPI

tseaver committed:
ci: test release build on PR branch

tseaver committed:
chore: configure qeumu for cross-compiles

tseaver committed:
fix: typo

tseaver committed:
ci: split linux / macos builds by arch

tseaver committed:
ci: s/arg/arch/

tseaver committed:
ci: remove running 'python-publish' workflow on push

Instead, run nightly, as well as on creation of a release.

Uploads to PyPI only occur during release runs.

tseaver committed:
ci: tweak required builds for presubmit
